Replace Reply Count

martindsq opened the pull request:
## Issues covered
#1246

## Description

Apart from replacing the reply count in the Home and Profile screens with "in discussion" or "Join the discussion" (and leaving the thread view almost the same), this PR solves several technical issues I found during the adventure:

- The reply count request we were using before didn't implement it in a way that if we could know the database was updated with new replies, that's another reason why most of the time the label displayed a wrong reply count. I changed the implementation so that we use the fetch request with the `@FetchRequest` annotation that enables us to detect those out of the box.
- I added two light relay requests to fetch new replies
- Added a code in the relay subscription model to limit the active subscriptions to an arbitrary number I chose based on testing in the real world in order to avoid complaining relays about too many REQs
- I disabled the reply count of root notes stacked on top of notes (we use this in the Profile and Home feed to show replies) in order to avoid fetching replies from the db in a case where the card is hidden visually.

I tried to do a fetch request that just returns the count of replies in core data using the `@FetchRequest` annotation so that we detect a change and update the reply count but didn't find a way, it seems that it is not supported in core data + swiftui. It's a shame. The best option I found was doing this very similar to how we were doing it before, fetch a collection of replies and use the count and filter the known follows in order to compute the reply count and collection of avatars. Let me know if you have a better idea, I would be happy to find a better solution.

And finally, please, let me know if you find this less performant compared to staging. I didn't, but we should keep our eyes open because we don't want a regression in terms of performance.

## How to test
1. Navigate to Home Feed
2. Look for a note with replies

## Screenshots/Video

![Simulator Screenshot - iPhone 15 - 2024-07-02 at 17 13 37](https://github.com/planetary-social/nos/assets/1703242/544ccaef-b46c-4946-817f-8a9e00281bd0)
![Simulator Screenshot - iPhone 15 - 2024-07-02 at 17 13 14](https://github.com/planetary-social/nos/assets/1703242/033ff266-a11f-48ad-b786-6166af4a9b33)

martindsq commented on Nos/Service/Relay/RelaySubscriptionManager.swift:
```diff
@@ -166,11 +166,40 @@ actor RelaySubscriptionManagerActor: RelaySubscriptionManager {
+        var waitingOneTimeSubscriptions = [RelaySubscription]()
+        var activeSubscriptionsCount = [URL: Int]()
+
+        all.forEach { relaySubscription in
```
This is a little refactor to avoid iterating over `all` more than one time, the code is not nicer, but it should be a bit more performant

martindsq commented on Nos/Views/ReplyCountLabel.swift:
```diff
@@ -0,0 +1,154 @@
+        !replies.isEmpty
+    }
+
+    @MainActor
```
This is because socialGraph depends on the main queue. Maybe there are other solutions that avoid this, but given that we expect a low number of replies, I guess this solution doesn't have an impact on performance.

mplorentz replied:
In general in the world of async/await it's fine to call long-running `async` functions on the MainActor. Because every time we `await` something on the MainActor it doesn't block the thread - it yields the thread and the MainActor is free to do other things while the awaited function (in this case running on the SocialGraphCache actor) does work elsewhere.

mplorentz commented on Nos/Views/NoteButton.swift:
```diff
@@ -13,8 +13,8 @@ struct NoteButton: View {
     var hideOutOfNetwork: Bool
-    var showReplyCount: Bool
-    var displayRootMessage: Bool 
+    var replyCount: ReplyCount?
```
Also I would find it more expressive to have `.none` be a style instead of using `nil`.

joshuatbrown replied:
I made a suggestion around the same time you did -- I think we're aligned on direction and just need a good name. I suggested `repliesDisplayType` since I prefer not to have `count` in the name. I'm open to other options, though.

martindsq replied:
Good suggestions! I renamed it to `RepliesDisplayType` and added a `.none` case

mplorentz commented on Nos/Views/NoteCard.swift:
```diff
@@ -136,16 +115,6 @@ struct NoteCard: View {
             Task { await note.loadAttributedContent() }
         }
         .background(LinearGradient.cardBackground)
-        .task {
```
I love that we are factoring so much of this out!

mplorentz commented on Nos/Views/ReplyCountLabel.swift:
```diff
@@ -0,0 +1,154 @@
+import Dependencies
+import SwiftUI
+
+struct ReplyCountLabel: View {
```
I opened this file up hoping for a SwiftUI preview. Can you add one? Now that we have them working again I think we should get back to our convention of adding a preview for every view that isn't trivial.

(I tried previewing NoteButton and NoteCard as well but I couldn't see this view rendered there either.)

martindsq replied:
Previews are still not working for me. I need to downgrade, but I can only do that once I'm back home due to network limitations (downloading Xcode). I've opened a ticket [here](https://github.com/planetary-social/nos/issues/1297) to ensure I don't forget about this.

mplorentz replied:
Did you try building for testing before rendering the preview? I thought they worked on the newer versions of xcode with that caveat.

martindsq replied:
Yes, but they don't work for me at the moment unfortunately.

mplorentz commented on Nos/Views/NoteCard.swift:
```diff
@@ -181,9 +150,7 @@ struct NoteCard_Previews: PreviewProvider {
         .environment(\.managedObjectContext, previewData.previewContext)
-        .environmentObject(previewData.relayService)
         .environmentObject(previewData.router)
-        .environment(previewData.currentUser)
```
This causes the preview to crash for me. Can you explain why you removed it?

martindsq replied:
I removed it while removing a currentUser dependency that I added while working on it, I'll bring it back

joshuatbrown commented on Nos/Models/ReplyCount.swift:
```diff
@@ -0,0 +1,8 @@
+/// Kind of counter (used when displaying a Note)
+enum ReplyCount {
```
I'd like to propose a different name for this enum but I'm struggling. `Count` only applies to one of the cases, so a broader term would be nice. `RepliesDisplayType`, maybe?

martindsq replied:
Renamed to RepliesDisplayType

mplorentz commented on Nos/Views/ReplyCountLabel.swift:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    func subscribeToReplies() {
+        Task(priority: .background) {
```
I think we should bump this up to `.userInitiated`. I've seen the OS hold off on scheduling `.background` priority tasks for several seconds or even minutes if the CPU is constantly busy. In this case we are loading data for the UI in response to the user scrolling so it makes sense to fire these off after the CPU is done handling tasks at the highest priority level.

joshuatbrown commented on Nos/Service/Relay/Filter.swift:
```diff
@@ -14,7 +14,10 @@ struct Filter: Hashable, Identifiable {
+    /// This filter should keep the subscription open.
+    var subscribe: Bool
+
     init(
```
This has a lot of parameters, but I generally like to have doc comments for all of them on non-private `init`s (and really anything that's non-private). One of the things I'd really like to have here in the docs would be the sentence "Defaults to `true`." for the `subscribe` parameter. "Defaults to ..." comments for the others would be great, too.

martindsq replied:
That's a good idea. I didn't add any of them but I think I know what each one does in order to write meaningful docs for all of them.

joshuatbrown commented on Nos/Service/Relay/RelaySubscriptionManager.swift:
```diff
@@ -166,15 +169,51 @@ actor RelaySubscriptionManagerActor: RelaySubscriptionManager {
-        if all.count > active.count {
-            Log.debug("\(all.count - active.count) subscriptions waiting in queue.")
+        #if DEBUG
+        let allCount = all.count
+        let activeCount = active.count
```
I think I preferred the other way where we didn't have variables for these since this adds two additional lines of code without improving readability.

martindsq replied:
it does not improve readability. However, active is a computed property that iterates over all and filter all those who are active. And the old code used it twice just to print to the console. This way, while a bit weird, doesn't has that performance penalty.

joshuatbrown commented on Nos/Views/AuthorStoryView.swift:
```diff
@@ -192,23 +192,9 @@ fileprivate struct BottomOverlay: View {
-                        .foregroundColor(Color.secondaryTxt)
-                }
-            }
+            ReplyCountLabel(replyCount: .discussion, for: note)
```
I think I'd prefer this:

```suggestion
            RepliesLabel(for: note, displayType: .discussion)
```

To me, that's a little easier to read and understand.

joshuatbrown commented on Nos/Views/NoteButton.swift:
```diff
@@ -13,8 +13,8 @@ struct NoteButton: View {
     var hideOutOfNetwork: Bool
-    var showReplyCount: Bool
-    var displayRootMessage: Bool 
+    var replyCount: ReplyCount?
```
```suggestion
    var repliesDisplayType: RepliesDisplayType?
```

Obviously I'd also update the `init` parameter to match.

joshuatbrown reviewed (changes requested):
A few naming suggestions -- this was just a review of the code. I didn't test it.

mplorentz reviewed (changes requested):
Hey Martin, this is great work. I'm really excited for this feature as it makes scrolling the feed feel much more alive. I left a lot of feedback because I think it's important we get this right. 
- Performance when scrolling Home and Profile feeds is great. However I'm seeing a major performance issue in large threads. For example it took almost a full minute before any replies showed for `note1yd4zd95ys6534xgtqyjllu2p7p6mh3ta3ujymwmray65hdeqzxeswqkd72`. From the logs I could see that there were 100+ subscriptions in the queue and they were very slowly cycling through. I think maybe we need to avoid opening subscriptions for every NoteCard in the ThreadView. I also get the feeling that subscriptions aren't getting cleared out of the `RelaySubscriptionManager.active` array fast enough, although I didn't confirm that.
- We lost the StackedAvatars view when ReplyCount is .count. I miss it and don't see that removal specified in the ticket anywhere. Can we bring it back?
- When I'm scrolling in a large thread after the note cards load in (for the first time, when it's downloading all the replies) I can consistently get the app to freeze up. Not sure why but I haven't seen this on any other branches.
- The "Join the discussion" button doesn't work at all in the stories interface (except on reposts)
- I see 504 build warnings on this branch that I don't see on `main`. No idea why.
- Design: given that the "Join the discusson" label uses a verb at the beginning implies that tapping it should do something - it should be a button. But it doesn't look like a button and it doesn't give feedback when tapped, and it doesn't really make me want to join the discussion lol. I think it would look better closer to the Figma design - if we moved the (white, not orange) reply button over to trailing edge of the "Join the discussion" text and made the text bold. Then both the text and the button should dim when tapped. Maybe we also move the reply button into the same place even when showing "in discussion". Just a suggestion but I think the design needs some more iterations.
- Design: I really like the avatars next to the "in discussion" label, but I often only see one even if the thread is huge. I think it would be better if we showed avatars even of people you don't follow. Maybe we prioritize people you do follow, but fall back to showing avatars of other folks. In fact I like this button so much that I could be convinced to get rid of the bland "Join the discussion" state entirely.

I'm fine with merging this without addressing my design feedback. Those might be better discussions to have with a larger group after they have played with it on staging. But if you want to tweak the design of the first iteration here a bit I'm fine with that too.

joshuatbrown commented on Nos/Service/Relay/Filter.swift:
```diff
@@ -14,7 +14,10 @@ struct Filter: Hashable, Identifiable {
     var until: Date?
-    
+
+    /// This filter should keep the subscription open.
```
Since this isn't a simple on/off thing, I wonder if it would be easier to understand if we have a two-case enum. Maybe something like `FilterType` with cases named something like `subscribeToAll` and `oneTime` (or `oneAndDone`? 😂 Just kidding but I love the rhyme). I also understand that this may be overkill and the Bool might be fine with the right documentation like Matt suggested.

joshuatbrown replied:
In thinking more about this, I'm leaning more towards a `FilterType` enum (or should it be `SubscriptionType`?), maybe with cases named `allNewEvents` and `allStoredEvents` or just `newEvents` and `storedEvents`. I'm probably not the best to name this since I don't fully understand it, but that's how Matt's comment reads to me.

mplorentz replied:
The cases would be more like `storedEvents` and `storedEventsAndNewEvents`. I'm fine with the bool because these aren't really two disjoint sets of events. Both cases will return stored events and the question is whether we want to keep the subscription open and stream in new events - which is a boolean question. But I'm also not opposed to an enum if you feel it's more expressive.

mplorentz replied:
Is part of the problem that "subscribe" is too generic of a verb? Maybe `streamNewEvents` or `observeNewEvents` would be better?

joshuatbrown replied:
I'm not 100% sure, but yeah, I think it's too generic for me. Aren't we always subscribing, technically? And this `Bool` just determines whether we keep the subscription open? So maybe something along those lines, or maybe `observeNewEvents`.

martindsq replied:
I'm changing it to `observeNewEvents` in RelaySubscription and `shouldKeepSubscriptionOpen` in Filter.

joshuatbrown commented on Nos/Service/Relay/RelayService.swift:
```diff
@@ -238,7 +267,13 @@ extension RelayService {
         }
         
-        return await subscribeToEvents(matching: Filter(eventIDs: [eventID], limit: 1))
+        return await subscribeToEvents(
```
I like `fetchEvents`=

martindsq replied:
I renamed it to `fetchEvents`

martindsq commented on Nos/Models/RelaySubscription.swift:
```diff
@@ -30,7 +30,7 @@ struct RelaySubscription: Identifiable, Hashable {
     
     /// Returns true if this is a "one-time" filter, where we are only looking for a single event
     var isOneTime: Bool {
```
I feel more comfortable renaming it to something else more meaningful for you. If I remove this property, the clients of RelaySubscription need to access the filter property every time and as a general rule they should not be dependent on the implementation of RelaySubscription, we need two hide that. Can you think about a better name? If you don't, maybe that's an indicator that we should remove .subscribe from Filter and move it to RelaySubscription.

mplorentz replied:
Oh I see. I think my brain thought this code was on `Filter`. In that case I think we should rename it `subscribe` or whatever the property is named on `Filter`.

martindsq replied:
See my reply in the other related discussion. Let's continue the discussion there.

martindsq commented on Nos/Service/Relay/RelaySubscriptionManager.swift:
```diff
@@ -37,6 +37,9 @@ actor RelaySubscriptionManagerActor: RelaySubscriptionManager {
     }
 
+    /// Limit of the number of active subscriptions in a single relay
+    private let queueLimit = 25
```
I could push the number to 25 at least when considering the recommended relays. There might be some relays out there that complain about that.

martindsq commented:
> * We lost the StackedAvatars view when ReplyCount is .count. I miss it and don't see that removal specified in the ticket anywhere. Can we bring it back?

I added it back. Plus, I'm filtering those who are not muted when looking for the avatars to avoid displaying someone the user doesn't want.

> * The "Join the discussion" button doesn't work at all in the stories interface (except on reposts)

Good catch, I removed the functionality there. I just added it back.

> * I see 504 build warnings on this branch that I don't see on `main`. No idea why.

They are all warnings about the localization enum we use. Relevant comment: https://planetary-app.slack.com/archives/C04PLR5QFQQ/p1720469420079109

> * Design: given that the "Join the discusson" label uses a verb at the beginning implies that tapping it should do something - it should be a button. But it doesn't look like a button and it doesn't give feedback when tapped, and it doesn't really make me want to join the discussion lol. I think it would look better closer to the Figma design - if we moved the (white, not orange) reply button over to trailing edge of the "Join the discussion" text and made the text bold. Then both the text and the button should dim when tapped. Maybe we also move the reply button into the same place even when showing "in discussion". Just a suggestion but I think the design needs some more iterations.
> * Design: I really like the avatars next to the "in discussion" label, but I often only see one even if the thread is huge. I think it would be better if we showed avatars even of people you don't follow. Maybe we prioritize people you do follow, but fall back to showing avatars of other folks. In fact I like this button so much that I could be convinced to get rid of the bland "Join the discussion" state entirely.

I find some your last comment on point, I also feel we are filtering why too much and filling the missing avatars with random avatars from the entire sets of repliers is not that bad. While I'm not totally convinced by the first one because I think we should not be moving the reply button from the right to the left preventing the user from habituating to it, I agree we should look for improving the design of the bottom row, its one of the core components if the UI, the users use it a lot. I think we should look for iterating the design in the next sprints.

martindsq commented:
Ok, I addressed all the feedback so far with the exception of the performance issues in threads and the previews thing (I have to downgrade Xcode). I think the priority would be understanding the performance issues and checking if we can do something more performant when displaying threads.

joshuatbrown commented on Nos/Models/RelaySubscription.swift:
```diff
@@ -28,9 +28,11 @@ struct RelaySubscription: Identifiable, Hashable {
+    /// Whether this RelaySubscription should subscribe to all new events
+    /// matching the filter (if true) or close once a relay has returned all
+    /// stored events (if false).
+    var observeNewEvents: Bool {
```
Maybe this is where my naming suggestion below should actually be used. Perhaps we could call this `keepAlive` or `keepOpen` or `persistentConnection` or something?

martindsq replied:
keepAlive or keepOpen should still cause the code to contain the `!`, which you complained to in the last comment and I agree with. So we have to think about a better name that communicates the opposite. What do you think about `closesAfterResponse`.? @joshuatbrown

joshuatbrown replied:
Yeah, I think that could work.

joshuatbrown commented on Nos/Views/NoteCard.swift:
```diff
@@ -96,14 +78,11 @@ struct NoteCard: View {
-                                        .font(.clarity(.medium, textStyle: .subheadline))
-                                        .foregroundColor(Color.secondaryTxt)
-                                }
+                            if repliesDisplayType != .displayNothing {
+                                RepliesLabel(
+                                    repliesDisplayType: repliesDisplayType,
+                                    for: note
+                                )
```
I get why we might want to check repliesDisplayType first, but couldn't RepliesLabel just return `EmptyView` if it's `.displayNothing`? This could would read more nicely for sure:

```suggestion
                            RepliesLabel(
                                repliesDisplayType: repliesDisplayType,
                                for: note
                            )
```

martindsq replied:
RepliesLabel already returns an `EmptyView`, I had to add that after receiving the feedback of stop using an optional but adding a .none case (that I had to change to .displayNothing because .none is already implemented in Swift. So far so good, but I don't like to do what you suggest here. Right now, the spacing of the HStack is zero, but if it were different than zero, having a view that the implementation contains an EmptyView adds an extra spacing. So, as a general rule, either communicate that with an optional, or do the checking twice, here and in RepliesLabel.

joshuatbrown reviewed (changes requested):
This has improved quite a bit since my last review. I'm still hoping we can make some naming and implementation changes, but the code is looking good overall.

martindsq commented on Nos/Service/Relay/Filter.swift:
```diff
@@ -4,39 +4,69 @@ import Foundation
+
+    /// Query string to use in the search parameter.
+    ///
+    /// Defaults to `nil`
```
I agree with you here. I did it that way because I thought you were suggesting that, but I do believe it should be in the init, I'll update this.

martindsq commented on Nos/Service/Relay/Filter.swift:
```diff
@@ -4,39 +4,69 @@ import Foundation
+    /// updates or should close after receiving a response from the relay.
+    ///
+    /// Defaults to `false`
+    var shouldKeepSubscriptionOpen: Bool
```
I don't think the Filter should control this. But I don't want to keep increasing the scope of the PR. I'm fine with changing it to `keepsSuscriptionOpen` to match the Swift suggestion of  "properties should read as assertions about the receiver".

joshuatbrown replied:
Makes sense -- I'm good with `keepSubscriptionOpen` (as opposed to `keepsSubscriptionOpen`).

joshuatbrown replied:
Actually I don't know. Maybe `keepsSubscriptionOpen` is fine, too. I'll leave that up to you.

martindsq commented on Nos/Service/Relay/RelayService.swift:
```diff
@@ -273,7 +303,7 @@ extension RelayService {
         if let subID = responseArray[1] as? String,
             let subscription = await subscriptionManager.subscription(from: subID),
-            subscription.isOneTime {
+            !subscription.observeNewEvents {
```
I'm replying below

joshuatbrown commented on Nos/Views/RepliesLabel.swift:
```diff
@@ -0,0 +1,173 @@
+
+        let noteIdentifier = note.identifier ?? ""
+
+        let format = """
+            SUBQUERY(
+                eventReferences,
+                $e,
+                $e.referencedEvent.identifier = %@ AND
+                    ($e.marker = 'reply' OR $e.marker = 'root')
+            ).@count > 0
+        """
+
+        let predicate = NSCompoundPredicate(
+            andPredicateWithSubpredicates: [
+                NSPredicate(format: "kind = 1"),
+                NSPredicate(
+                    format: format,
+                    noteIdentifier,
+                    noteIdentifier
+                ),
+                NSPredicate(format: "deletedOn.@count = 0"),
+                NSPredicate(format: "author.muted = false")
+            ]
+        )
+
+        let fetchRequest = Event.fetchRequest()
+        fetchRequest.includesPendingChanges = false
+        fetchRequest.includesSubentities = false
+        fetchRequest.sortDescriptors = [
+            NSSortDescriptor(keyPath: \Event.identifier, ascending: true)
+        ]
+
+        fetchRequest.predicate = predicate
+        fetchRequest.relationshipKeyPathsForPrefetching = ["author"]
+        _replies = FetchRequest(fetchRequest: fetchRequest)
```
I think a lot of this could move out of here, right? And this could be:

```suggestion
        _replies = Event.replies(to: note)
```

I see we already have an `Event.allReplies(to:)` but I'm assuming that's not exactly what you need here. If so, we can use that! If not, I'd just put another class func there.

joshuatbrown reviewed (changes requested):
I think we're close on this, and I promise I'll stop nitpicking! These are small things -- overall this is looking awesome!

martindsq commented:
> * Performance when scrolling Home and Profile feeds is great. However I'm seeing a major performance issue in large threads. For example it took almost a full minute before any replies showed for `note1yd4zd95ys6534xgtqyjllu2p7p6mh3ta3ujymwmray65hdeqzxeswqkd72`. From the logs I could see that there were 100+ subscriptions in the queue and they were very slowly cycling through. I think maybe we need to avoid opening subscriptions for every NoteCard in the ThreadView. I also get the feeling that subscriptions aren't getting cleared out of the `RelaySubscriptionManager.active` array fast enough, although I didn't confirm that.
> * When I'm scrolling in a large thread after the note cards load in (for the first time, when it's downloading all the replies) I can consistently get the app to freeze up. Not sure why but I haven't seen this on any other branches.

I moved the subscription to relays code to NoteButton and made it optional. We only need that for HomeFeed and Profile Feed because on the Replies View, we already fetch all replies to the note the user is seeing, so we don't need to do it again nor for every reply in the discussion.

martindsq commented:
> I think we're close on this, and I promise I'll stop nitpicking! These are small things -- overall this is looking awesome!

Thanks for the thorough review. I believe I have addressed all the points.

mplorentz commented on Nos/Service/Relay/RelaySubscriptionManager.swift:
```diff
@@ -184,16 +183,12 @@ actor RelaySubscriptionManagerActor: RelaySubscriptionManager {
-            } else {
-                waitingLongSubscriptions.append(relaySubscription)
             }
+            waitingSubscriptions.append(relaySubscription)
```
shouldn't this be inside an `else` block? The way this reads now it looks like we will add active subscriptions to the waiting queue, and then start them again.

mplorentz commented on Nos/Service/Relay/RelayService.swift:
```diff
@@ -188,7 +188,9 @@ extension RelayService {
-    func requestAReplyFromAnyone(for eventID: String?) async -> SubscriptionCancellable {
+    func requestReplyFromAnyone(
+        for eventID: String?
+    ) async -> SubscriptionCancellable {
```
👍 my MO has always been to put the function definition on a single line if it fits on a single line, but I'm not sure that's written down anywhere.

joshuatbrown replied:
That's how I do it, too. I'm not sure where we'd write this -- maybe a style guide? I don't think we have one, so maybe it's time to start.

mplorentz commented on Nos/Views/NoteButton.swift:
```diff
@@ -131,6 +158,34 @@ struct NoteButton: View {
+            relaySubscriptions.removeAll()
+            relaySubscriptions.append(
+                await relayService.requestReplyFromAnyone(
+                    for: note.identifier
```
🤔 this code appears to be working, but I'm pretty sure we want to use `displayedNote` here and in the other call to relayService. In the case of a repost `note` will be the kind 6 and `displayedNote` will be the kind 1 we want to fetch replies for.
```suggestion
                    for: displayedNote.identifier
```

martindsq replied:
You are right. I missed that after moving the code to NoteButton.

mplorentz commented on Nos/Views/RepliesLabel.swift:
```diff
@@ -0,0 +1,113 @@
+import Dependencies
+import SwiftUI
+
+struct RepliesLabel: View {
```
```suggestion
/// Displays a summary of the reply activity for the given note, in different styles enumerated in `RepliesDisplayType`.
struct RepliesLabel: View {
```

mplorentz commented on Nos/Views/NoteCard.swift:
```diff
@@ -96,14 +78,11 @@ struct NoteCard: View {
-                                        .foregroundColor(Color.secondaryTxt)
-                                }
+                            if repliesDisplayType != .displayNothing {
+                                RepliesLabel(
```
Can we put this inside it's own `Button`? It's really nice to see it give feedback when tapped in the `StoryNoteView` and it would be nice to get it here as well.

mplorentz reviewed (changes requested):
The performance in large threads is fixed! And I'm not seeing the freeze I was seeing earlier anymore.

I'm requesting a few changes - the one in RelaySubscriptionManager is pretty major but the rest are minor. I'd like to give it one more review when you have completed them though. Your code looks great but I am paranoid about correctness and performance in the RelayService because it is so brittle.

mplorentz commented:
Not sure how I posted two reviews back to to back 🤔 that should have all been one review.

martindsq commented on Nos/Service/Relay/Filter.swift:
```diff
@@ -113,8 +150,7 @@ struct Filter: Hashable, Identifiable {
             since?.timeIntervalSince1970.description ?? "nil",
-            until?.timeIntervalSince1970.description ?? "nil",
-            inNetwork.description,
+            until?.timeIntervalSince1970.description ?? "nil"
```
I did it the other way around: the hash function uses just the id now.

joshuatbrown replied:
What led you to do it the other way around? It looks to me like it would have been simpler to just add a `hasher.combine(keepSubscriptionOpen)` line up there than to have to get its description down here. And the other more complex lines here could be removed, too, which would be nice. But I didn't try it and might be missing something.

martindsq replied:
I saw it implemented that way in https://medium.com/@saradrada/hashable-vs-identifiable-protocols-in-swift-71bd310c50cb with the hash function dependent on the id property. But I'll update it to match your preference.

joshuatbrown commented on Nos/Models/CoreData/Event+CoreDataClass.swift:
```diff
@@ -324,7 +324,48 @@ public class Event: NosManagedObject, VerifiableEvent {
+    ///
+    /// Intented to be used primarily to compute the number of replies and for
+    /// building a set of author avatars.
+    @nonobjc public class func replies(to noteIdentifier: String) -> FetchRequest<Event> {
```
The method that was removed had `noteID: RawEventID` as a parameter -- is there a reason to do it differently here? I think ideally we'd be using the `RawEventID` wherever we can to signify that this is an event id and not just any String. I could be missing something, though.

```suggestion
    @nonobjc public class func replies(to noteID: RawEventID) -> FetchRequest<Event> {
```

martindsq replied:
I renamed it to noteID and changed the String to RawEventID.

joshuatbrown commented on Nos/Service/Relay/RelayService.swift:
```diff
@@ -188,6 +188,40 @@ extension RelayService {
         )
     }
     
+    func requestReplyFromAnyone(for eventID: String?) async -> SubscriptionCancellable {
```
Maybe this?
```suggestion
    func requestReplyFromAnyone(for eventID: RawEventID?) async -> SubscriptionCancellable {
```

joshuatbrown commented on Nos/Service/Relay/RelayService.swift:
```diff
@@ -188,6 +188,40 @@ extension RelayService {
+    /// - Parameter limit: Maximum number of replies to ask for. Defaults to
+    /// nil (no limit).
+    func requestRepliesFromFollows(
+        for eventID: String?,
```
```suggestion
        for eventID: RawEventID?,
```

joshuatbrown commented on Nos/Views/NoteButton.swift:
```diff
@@ -13,20 +13,46 @@ struct NoteButton: View {
+    /// Initializes a NoteButton object.
+    ///
+    /// - Parameter note: Note event to display.
+    /// - Parameter style: Card style. Defaults to `comapact`.
```
```suggestion
    /// - Parameter style: Card style. Defaults to `.compact`.
```

martindsq replied:
I also updated `displayNothing` to `.displayNothing` below to match your style of starting with the dot

joshuatbrown reviewed (commented):
This has improved quite a bit since it was opened. I have a few more comments so we can be more clear and consistent. Thanks for adding all the docs and tackling all this work!

martindsq commented:
@joshuatbrown it is ready to be reviewed again

mplorentz reviewed (approved):
I just spent a long time testing this and it looks solid to me. Nice work @martindsq. This is a big change but it makes the app feel much more alive. The one additional change I suggested is crucial but I'm going to go ahead and approve so you can merge after @joshuatbrown approves too.

I also made some changes to take advantage of the new `keepSubscriptionOpen` behavior and close some subscriptions faster in #1303.

joshuatbrown reviewed (approved):
Looks great and works well! Awesome work on this! My approval is also conditional on you addressing Matt's suggested changes, but otherwise this is awesome!

martindsq commented on Nos/Service/Relay/RelaySubscriptionManager.swift:
```diff
@@ -106,7 +109,7 @@ actor RelaySubscriptionManagerActor: RelaySubscriptionManager {
         var staleSubscriptions = [RelaySubscription]()
         for subscription in active {
-            if subscription.isOneTime, 
+            if !subscription.closesAfterResponse, 
```
Yes, you're right, I missed this after renaming it to subscribes (which did the opposite thing) and then to closesAfterResponse.
